### PR TITLE
[WIP] OCPBUGS-988: Let rpm-ostree trust and pull from internal registry, especially during "firstboot" 

### PR DESCRIPTION
**Pull request by jkyros**
## This is an imperfect attempt at seamless internal registry integration for layering to frame/demonstrate the scope of the issue, this should not merge without refinement. 

Ideally a user would be able to override `osImageURL` with an image from the cluster's internal registry seamlessly without having to do any sort of setup(if the internal registry is present), but today rpm-ostree doesn't have the credentials (or certificate trust trust, in the case of firstboot) to be able to do that. 

### We have (at least) three obstacles: 

1. The MCO/rpm-ostree/skopeo does not have a pull secret for the internal registry  
2. The internal registry by default has a self-signed cert that it sneaks in during live operation **as another privileged daemonset** `oc -n openshift-image-registry describe ds/node-ca` (https://github.com/openshift/cluster-image-registry-operator/blob/e038fc4a9f87d59dc1f372ebd253d70d08a20b0d/pkg/resource/podtemplatespec.go#L489), but that isn't there yet during firstboot 
3. The internal registry does not listen on the host network, and during firstboot we don't have an SDN so we can't access the cluster's pod/service network (this is okay during firstboot during cluster install because the internal registry doesn't exist, but this is *not* okay during scaling up a node _after you've applied an image from the internal registry_)

### This PR solves those problems (imperfectly) by: 
1. Adding a new pull-only serviceaccount for the MCO, adding it to machineconfig, and then having the MCD merge it together with the kubelet/global pull secret into `/run/ostree/auth.json`
2. Adding an additional trust bundle via `MachineConfig`, so firstboot rpm-ostree will trust the certificate, and also allowing the MCO apply/update CA trust without a reboot
3. Allowing the MCD to access the image registry via its default route (when exposed) and to use the images there interchangeably whether they come from the "internal" or the "external" registry hostname. 

### This also has the following undesirable side-effect: 

We have to roll to a new rebootless machine-config shortly after cluster bringup is finished (once the registry shows up, `MachineConfig` changes because the certificates are present). 
```
[jkyros@jkyros-t590 cluster]$ oc get mc
NAME                                               GENERATEDBYCONTROLLER                      IGNITIONVERSION   AGE
00-master                                          6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
00-worker                                          6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
01-master-container-runtime                        6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
01-master-kubelet                                  6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
01-worker-container-runtime                        6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
01-worker-kubelet                                  6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
99-master-generated-registries                     6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
99-master-ssh                                                                                 3.2.0             120m
99-worker-generated-registries                     6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m
99-worker-ssh                                                                                 3.2.0             120m
rendered-master-95fa64f22b4e8eac9890700cd55c6125   6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             104m
this --> rendered-master-9688bd6425e7072d50d251c03490a7f4   6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m 
this --> rendered-worker-537b11b404a3831558d6688a4f4a1536   6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             118m 
rendered-worker-d7c0288aae69396a444e768bb0873b6e   6def4035bf2da882da2ad4937196dd806341daa6   3.2.0             104m
```

### "You said it works! I want to see it work!"
After your cluster is up...

1.  Get your base image
```
$ oc adm release info --image-for rhel-coreos-8
quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d16e3e31dd727f8481bf468dbc0064804b33b3d5ae5dd932040a9f3e9bf37329
```
4.  Do a new in-cluster build 
```
$ oc project openshift-machine-config-operator
$ oc new-build -D $'FROM quay.io/openshift-release-dev/ocp-v4.0-art-dev@sha256:d16e3e31dd727f8481bf468dbc0064804b33b3d5ae5dd932040a9f3e9bf37329 \nRUN touch /etc/test-file-1' --to in-cluster-image
```
5.  Get the tag when it's done
```
$ oc get imagetag in-cluster-image:latest -o jsonpath="{.status.items[0].dockerImageReference}"
image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/in-cluster-image@sha256:c7c80d052b7b6a40491744ca0e38ba36a8bca67874d84848d048aa264283a848
```
6. Apply a machineconfig with it
```
apiVersion: machineconfiguration.openshift.io/v1
kind: MachineConfig
metadata:
  labels:
    machineconfiguration.openshift.io/role: worker
  name: 99-jkyros-external-image-worker 
spec:
  osImageURL: "image-registry.openshift-image-registry.svc:5000/openshift-machine-config-operator/in-cluster-image@sha256:c7c80d052b7b6a40491744ca0e38ba36a8bca67874d84848d048aa264283a848"

```
7. Wait for the MCO to complete at least one node

8. Expose the internal registry's default route before you can scale up a new node (because otherwise our pre-kubelet firstboot pull can't get there)
```
$ oc patch configs.imageregistry.operator.openshift.io/cluster --patch '{"spec":{"defaultRoute":true}}' --type=merge
```
9.  Scale up a new machine to prove firstboot works

```
$ oc scale --replicas=2 -n openshift-machine-api machineset jkyros-ocpbugs-988-17-r7dp4-worker-us-east-1b 
```

### Outstanding questions: 
- We always thought we might need to do an "in-cluster build" at install time, is this the time to think through what that might look like? 
- Wouldn't it be nice if all of the default certificates for all of our operators inherited from some shared trust source so we didn't have to piecemeal a trust bundle? :smile: (I would *love* to be ignorant here and find out this is somehow being handled by some other operator and we're just not wise to it yet) 

 
## TL;DR 
- internal registry trust isn't seamless, it should be
- I'd like to find a more elegant way to do this
- This PR is for illustration purposes, *it should not merge*

### CODE REVIEW

File: pkg/daemon/update.go [hunk 1]
```diff
@@ -2188,6 +2207,14 @@ func (dn *CoreOSDaemon) applyLayeredOSChanges(mcDiff machineConfigDiff, oldConfi
 		mcDiff.osUpdate = false
 	}
 
+	// TODO(jkyros): the registry name is different on the "outside" when we're in firstboot,
+	// but it's the same on the inside. I really should be checking the name + digest.
+	justTheImage := strings.TrimPrefix(dn.bootedOSImageURL, "image-registry.openshift-image-registry.svc:5000")
```
Comment (cgwalters) [inline pkg/daemon/update.go:2238]: Hmm.  I think we can actually *just* compare the digest, no?
